### DISCUSSION TITLE
Integrate VSG with VKVG

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1340] [Ideas]
From the https://github.com/jpbruyere/vkvg home page:

"vkvg is an open-source 2D graphics library written in C using [Vulkan](https://www.khronos.org/vulkan/) as backend."

I would like to see a vsg adapter so that vkvg can be used in vsg.

From my initial look at the sources this looks quite straightforward:

1. Create a `VkVgDevice` by (there are other "constructors" as well):
```
VkvgDevice vkvg_device_create_from_vk_multisample (VkInstance inst, VkPhysicalDevice phy, VkDevice vkdev, uint32_t qFamIdx, uint32_t qIndex, VkSampleCountFlags samples, bool deferredResolve);
```
2. Create a `VkvgSurface` by:
```
VkvgSurface vkvg_surface_create (VkvgDevice dev, uint32_t width, uint32_t height);
```
3. For each frame do cairo graphics ike drawing:
```
    VkvgContext ctx = vkvg_create(surf);
    vkvg_set_source_rgb(ctx,0.1,0.1,0.1);
    vkvg_paint(ctx);
    // More painting here
    vkvg_destroy(ctx);
```

So how does this map to vsg?

**Comment by robertosfield** [comment]
I'm not familiar with vkvg and as others haven't chipped in yet you might be the first in the VSG community to attempt this integration.  As I haven't used vkvg I can only ask questions and make general suggestions.

First thought, is vkvg designed to allow you to create your own vkInstance, vkDevice & vkSurface and have it adopts these and defer lifetime management of those to users of vkvg?  If so then getting the Vulkan handles from the the vsg::Instance, vsg::Device and vsg::Surface and passes this into the vkcg adapter classes.

On the rendering side (item 3. in your post) the way I'd integrate that with the scene graph would be subclass from vsg::Command and override the Command::record(..) method and have this call the vkvg code.
